### PR TITLE
FEAT: Support minicpm-4B

### PR DESCRIPTION
**Pull request by Jun-Howie**
Testing has been done locally and the model requires the following two dependencies to be added - datamodel_code_generator and jsonschema

### CODE REVIEW

File: .github/workflows/python.yaml [hunk 1]
```diff
@@ -136,6 +136,8 @@ jobs:
           pip install eva-decord
           pip install jj-pytorchvideo
           pip install qwen-vl-utils
+          pip install datamodel_code_generator
```
Comment (qinxuye) [inline .github/workflows/python.yaml:139]: Why does this model have this dependencies? Where can I refer to?

**Review by qinxuye — COMMENTED** [review]
Can you add this model to 

https://github.com/xorbitsai/inference/blob/d2a11960ff9d796efae4ccbbe60ad6d091422796/xinference/model/llm/vllm/core.py#L133

to support vllm?

File: .github/workflows/python.yaml (hunk 1 above)
Comment (Jun-Howie) [reply .github/workflows/python.yaml:139]: 2024-09-10 13:37:21,543 xinference.core.worker 688 ERROR    Failed to load model minicpm3-4b-1-0
Traceback (most recent call last):
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/worker.py", line 893, in launch_builtin_model
    await model_ref.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 231, in send
    return self._process_result_message(result)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 656, in send
    result = await self._run_coro(message.message_id, coro)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 367, in _run_coro
    return await coro
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 526, in xoscar.core._BaseActor.__on_receive__
    result = await result
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/model.py", line 309, in load
    self._model.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 755, in load
    super().load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 318, in load
    self._model, self._tokenizer = self._load_model(**kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 187, in _load_model
    tokenizer = AutoTokenizer.from_pretrained(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 877, in from_pretrained
    tokenizer_class = get_class_from_dynamic_module(class_ref, pretrained_model_name_or_path, **kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 502, in get_class_from_dynamic_module
    final_module = get_cached_module_file(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 327, in get_cached_module_file
    modules_needed = check_imports(resolved_module_file)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 182, in check_imports
    raise ImportError(
ImportError: [address=0.0.0.0:45191, pid=708] This modeling file requires the following packages that were not found in your environment: datamodel_code_generator, jsonschema. Run `pip install datamodel_code_generator jsonschema`
2024-09-10 13:37:21,599 xinference.core.worker 688 ERROR    [request 13db8fd6-6f36-11ef-b9cb-0242c0a84902] Leave launch_builtin_model, error: [address=0.0.0.0:45191, pid=708] This modeling file requires the following packages that were not found in your environment: datamodel_code_generator, jsonschema. Run `pip install datamodel_code_generator jsonschema`, elapsed time: 289 s
Traceback (most recent call last):
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/utils.py", line 69, in wrapped
    ret = await func(*args, **kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/worker.py", line 893, in launch_builtin_model
    await model_ref.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 231, in send
    return self._process_result_message(result)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 656, in send
    result = await self._run_coro(message.message_id, coro)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 367, in _run_coro
    return await coro
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 526, in xoscar.core._BaseActor.__on_receive__
    result = await result
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/model.py", line 309, in load
    self._model.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 755, in load
    super().load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 318, in load
    self._model, self._tokenizer = self._load_model(**kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 187, in _load_model
    tokenizer = AutoTokenizer.from_pretrained(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 877, in from_pretrained
    tokenizer_class = get_class_from_dynamic_module(class_ref, pretrained_model_name_or_path, **kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 502, in get_class_from_dynamic_module
    final_module = get_cached_module_file(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 327, in get_cached_module_file
    modules_needed = check_imports(resolved_module_file)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 182, in check_imports
    raise ImportError(
ImportError: [address=0.0.0.0:45191, pid=708] This modeling file requires the following packages that were not found in your environment: datamodel_code_generator, jsonschema. Run `pip install datamodel_code_generator jsonschema`
2024-09-10 13:37:21,617 xinference.api.restful_api 559 ERROR    [address=0.0.0.0:45191, pid=708] This modeling file requires the following packages that were not found in your environment: datamodel_code_generator, jsonschema. Run `pip install datamodel_code_generator jsonschema`
Traceback (most recent call last):
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/api/restful_api.py", line 876, in launch_model
    model_uid = await (await self._get_supervisor_ref()).launch_builtin_model(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 231, in send
    return self._process_result_message(result)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 656, in send
    result = await self._run_coro(message.message_id, coro)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 367, in _run_coro
    return await coro
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 526, in xoscar.core._BaseActor.__on_receive__
    result = await result
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/supervisor.py", line 1026, in launch_builtin_model
    await _launch_model()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/supervisor.py", line 990, in _launch_model
    await _launch_one_model(rep_model_uid)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/supervisor.py", line 969, in _launch_one_model
    await worker_ref.launch_builtin_model(
  File "xoscar/core.pyx", line 284, in __pyx_actor_method_wrapper
    async with lock:
  File "xoscar/core.pyx", line 287, in xoscar.core.__pyx_actor_method_wrapper
    result = await result
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/utils.py", line 69, in wrapped
    ret = await func(*args, **kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/worker.py", line 893, in launch_builtin_model
    await model_ref.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 231, in send
    return self._process_result_message(result)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/context.py", line 102, in _process_result_message
    raise message.as_instanceof_cause()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 656, in send
    result = await self._run_coro(message.message_id, coro)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/backends/pool.py", line 367, in _run_coro
    return await coro
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xoscar/api.py", line 384, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 558, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 520, in xoscar.core._BaseActor.__on_receive__
    async with self._lock:
  File "xoscar/core.pyx", line 521, in xoscar.core._BaseActor.__on_receive__
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 526, in xoscar.core._BaseActor.__on_receive__
    result = await result
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/core/model.py", line 309, in load
    self._model.load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 755, in load
    super().load()
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 318, in load
    self._model, self._tokenizer = self._load_model(**kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/xinference/model/llm/transformers/core.py", line 187, in _load_model
    tokenizer = AutoTokenizer.from_pretrained(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py", line 877, in from_pretrained
    tokenizer_class = get_class_from_dynamic_module(class_ref, pretrained_model_name_or_path, **kwargs)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 502, in get_class_from_dynamic_module
    final_module = get_cached_module_file(
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 327, in get_cached_module_file
    modules_needed = check_imports(resolved_module_file)
  File "/root/xinglin-data/xinf/lib/python3.10/site-packages/transformers/dynamic_module_utils.py", line 182, in check_imports
    raise ImportError(
ImportError: [address=0.0.0.0:45191, pid=708] This modeling file requires the following packages that were not found in your environment: datamodel_code_generator, jsonschema. Run `pip install datamodel_code_generator jsonschema`

**Comment by Jun-Howie** [comment]
> 您能否将此模型添加到
> 
> https://github.com/xorbitsai/inference/blob/d2a11960ff9d796efae4ccbbe60ad6d091422796/xinference/model/llm/vllm/core.py#L133
> 
> 支持 VLLM？

vllm is not officially supported yet，See the vllm branch maintained by minicpm

https://github.com/OpenBMB/vllm/tree/minicpm3
and
https://github.com/OpenBMB/vllm/blob/minicpm3/vllm/model_executor/models/minicpm3.py

**Review by qinxuye — APPROVED** [review]
LGTM

**Comment by linqingxu** [comment]
模型上线页面没有minicpm-4b模型

**Comment by Jun-Howie** [comment]
> 模型上线页面没有minicpm-4b模型

您是pip 安装的的xinference把，这个支持目前还没推送到pip源上，要等周五晚上发版

**Comment by linqingxu** [comment]
> > 模型上线页面没有minicpm-4b模型
> 
> 您是pip 安装的的xinference把，这个支持目前还没推送到pip源上，要等周五晚上发版

是用docker部署的，昨天pull的nightly-main镜像里找不到minicpm3-4b

**Comment by Jun-Howie** [comment]
> > > 模型上线页面没有minicpm-4b模型
> > 
> > 
> > 您是pip 安装的的xinference把，这个支持目前还没推送到pip源上，要等周五晚上发版
> 
> 是用docker部署的，昨天pull的nightly-main镜像里找不到minicpm3-4b

我拉了最新的nightly-main镜像 。是有这个模型的！

image

**Comment by linqingxu** [comment]
> > > > 模型上线页面没有minicpm-4b模型
> > > 
> > > 
> > > 您是pip 安装的的xinference把，这个支持目前还没推送到pip源上，要等周五晚上发版
> > 
> > 
> > 是用docker部署的，昨天pull的nightly-main镜像里找不到minicpm3-4b
> 
> 我拉了最新的nightly-main镜像 。是有这个模型的！
> 
> image

又试了一次，现在有了，请问能支持vllm推理吗，我看minicpm是有支持vllm的（得安装他们的分支）

**Comment by Jun-Howie** [comment]
> > > > > 模型上线页面没有minicpm-4b模型
> > > > 
> > > > 
> > > > 您是pip 安装的的xinference把，这个支持目前还没推送到pip源上，要等周五晚上发版
> > > 
> > > 
> > > 是用docker部署的，昨天pull的nightly-main镜像里找不到minicpm3-4b
> > 
> > 
> > 我拉了最新的nightly-main镜像 。是有这个模型的！
> > image
> 
> 又试了一次，现在有了，请问能支持vllm推理吗，我看minicpm是有支持vllm的（得安装他们的分支）

目前mincpm的vllm分支还没有合并到vllm主分支上，（所有暂时还不支持）你可以根据minicpm的文档单独进行编译

**Comment by haoguangxie** [comment]
支持sglang推理吗？我看minicpm是有支持sglang的

**Comment by qinxuye** [comment]
sglang 添加模型只需要添加模型名到列表：

https://github.com/xorbitsai/inference/blob/325626f2a46ddfcf6fc126cf77d1ab482f647edd/xinference/model/llm/sglang/core.py#L74

欢迎贡献  PR。

**Pull request merged**